packages/compass-e2e-tests/helpers/test-runner-context.ts

```diff
@@ -2,145 +2,312 @@ import {
-// compass-web.
-export const SKIP_COMPASS_DESKTOP_COMPILE =
-  process.argv.includes('--no-compile') && !TEST_COMPASS_WEB;
-// Skip this step if you want to run tests against your own compilation (e.g, a
-// dev build or a build running in watch mode that autorecompiles). Also no need
-// to recompile when testing compass-web.
-export const SKIP_NATIVE_MODULE_REBUILD =
-  process.argv.includes('--no-native-modules') && !TEST_COMPASS_WEB;
-export const DISABLE_START_STOP = process.argv.includes('--disable-start-stop');
-export const MOCHA_BAIL = process.argv.includes('--bail');
-
-export const COMPASS_WEB_BROWSER_NAME =
-  process.env.COMPASS_WEB_BROWSER_NAME ?? 'chrome';
-// https://webdriver.io/docs/driverbinaries/
-//
-// If you leave out browserVersion it will try and find the browser binary on
-// your system. If you specify it it will download that version. The main
-// limitation then is that 'latest' is the only 'semantic' version that is
-// supported for Firefox.
-// https://github.com/puppeteer/puppeteer/blob/ab5d4ac60200d1cea5bcd4910f9ccb323128e79a/packages/browsers/src/browser-data/browser-data.ts#L66
-//
-// Alternatively we can download it ourselves and specify the path to the binary
-// or we can even start and stop chromedriver/geckodriver manually.
-//
-// NOTE: The version of chromedriver or geckodriver in play might also be
-// relevant.
-export const COMPASS_WEB_BROWSER_VERSION =
-  process.env.COMPASS_WEB_BROWSER_VERSION === 'unset'
-    ? undefined
-    : process.env.BROWSER_VERSION ?? 'latest';
-export const COMPASS_WEB_SANDBOX_URL = 'http://localhost:7777';
-
-const MONGODB_TESTSERVER_VERSION =
+const argv = yargs(hideBin(process.argv))
+  .scriptName('compass-e2e-tests')
+  .env('COMPASS_E2E')
+  .detectLocale(false)
+  .version(false)
+  .strict();
+
+const parsedArgs = buildCommonArgs(argv)
+  .command(
+    ['$0', 'desktop'],
+    'Run e2e tests against Compass desktop',
+    buildDesktopArgs,
+    () => {
+      testEnv = 'desktop';
+    }
+  )
+  .command('web', 'Run e2e tests against Compass web', buildWebArgs, () => {
+    testEnv = 'web';
+  })
+  .parse();
+
+type BuilderCallbackParsedArgs<A extends (...args: any[]) => Argv<any>> =
+  ReturnType<ReturnType<A>['parseSync']>;
+
+type CommonParsedArgs = BuilderCallbackParsedArgs<typeof buildCommonArgs>;
+
+type DesktopParsedArgs = CommonParsedArgs &
+  BuilderCallbackParsedArgs<typeof buildDesktopArgs>;
+
+type WebParsedArgs = CommonParsedArgs &
+  BuilderCallbackParsedArgs<typeof buildWebArgs>;
+
+if (!testEnv) {
+  throw new Error('Test env was not selected');
+}
+
+if ('then' in parsedArgs && typeof parsedArgs.then === 'function') {
+  throw new Error('Async args parser is not allowed');
+}
+
+export const context = parsedArgs as CommonParsedArgs &
+  Partial<DesktopParsedArgs & WebParsedArgs>;
+
+export function isTestingDesktop(ctx = context): ctx is DesktopParsedArgs {
+  return testEnv === 'desktop';
```

Comment:
We can explicitly check for every desktop specific command argument here as defined above, but yargs with `strict()` guarantees that this assertion is correct if we managed to set `testEnv` to `desktop` (same assumption can be made for web)

Reply:
In this specific case it's unlikely that the assertion will be broken, but I wonder if there is a way to make the assertion less disconnected from the code that actually ensures this is valid

Reply:
I was thinking something like a map of the testEnv -> argsParser, which would be used to in both pieces of code. but looking a bit deeper into how the types are build, it's probably more complicated to implement that and probably not worth it given that the very small risk.